src/dpdkstat.c

```diff
@@ -0,0 +1,775 @@
+  if (err)
+    ERROR("dpdkstat: %s : error %d in shm_init()", __func__, err);
+
+  /* If the XML config() function has been run, dont re-initialize defaults */
+  if(!g_configured) {
+    dpdk_config_init_default();
+  }
+
+  plugin_register_complex_read (NULL, "dpdkstat", dpdk_read,
+                                g_configuration->interval, NULL);
+  return 0;
+}
+
+static int dpdk_helper_exit(int reset)
+{
+  g_configuration->helper_status = DPDK_HELPER_GRACEFUL_QUIT;
+  if (reset) {
+    g_configuration->eal_initialized = 0;
+    g_configuration->num_ports = 0;
+    memset(&g_configuration->xstats, 0, g_configuration->num_xstats* sizeof(struct rte_eth_xstats));
+    g_configuration->num_xstats = 0;
+    for (int i = 0; i < RTE_MAX_ETHPORTS; i++)
+      g_configuration->num_stats_in_port[i] = 0;
+  }
+  close(g_configuration->helper_pipes[1]);
+  int err = kill(g_configuration->helper_pid, SIGKILL);
+  if (err) {
+    ERROR("dpdkstat: error sending kill to helper: %s\n", strerror(errno));
+  }
+
+  return 0;
+}
+
+static int dpdk_helper_spawn(enum DPDK_HELPER_ACTION action)
+{
+  g_configuration->eal_initialized = 0;
+  g_configuration->helper_action = action;
+  /*
+   * Create a pipe for helper stdout back to collectd. This is necessary for
+   * logging EAL failures, as rte_eal_init() calls rte_panic().
+   */
+  if (g_configuration->helper_pipes[1]) {
+    DEBUG("dpdkstat: collectd closing helper pipe %d\n",
+          g_configuration->helper_pipes[1]);
+  } else {
+    DEBUG("dpdkstat: collectd helper pipe %d, not closing\n",
+          g_configuration->helper_pipes[1]);
+  }
+  if(pipe(g_configuration->helper_pipes) != 0) {
+    DEBUG("dpdkstat: Could not create helper pipe: %s\n", strerror(errno));
+    return -1;
+  }
+
+  int pipe0_flags = fcntl(g_configuration->helper_pipes[0], F_GETFL, 0);
+  int pipe1_flags = fcntl(g_configuration->helper_pipes[1], F_GETFL, 0);
+  if (pipe0_flags == -1 || pipe1_flags == -1) {
+    ERROR("dpdkstat: error setting up pipe flags: %s\n", strerror(errno));
+  }
+  int  pipe0_err = fcntl(g_configuration->helper_pipes[0], F_SETFL, pipe1_flags
+                         | O_NONBLOCK);
+  int  pipe1_err = fcntl(g_configuration->helper_pipes[1], F_SETFL, pipe0_flags
+                         | O_NONBLOCK);
+  if (pipe0_err == -1 || pipe1_err == -1) {
+    ERROR("dpdkstat: error setting up pipes: %s\n", strerror(errno));
+  }
+
+  pid_t pid = fork();
+  if (pid > 0) {
+    close(g_configuration->helper_pipes[1]);
+    g_configuration->helper_pid = pid;
+    DEBUG("dpdkstat: helper pid %u\n", g_configuration->helper_pid);
+    /* Kick helper once its alive to have it start processing */
+    sem_post(&g_configuration->sema_helper_get_stats);
+  } else if (pid == 0) {
+    /* Replace stdout with a pipe to collectd. */
+    close(g_configuration->helper_pipes[0]);
+    close(STDOUT_FILENO);
+    dup2(g_configuration->helper_pipes[1], STDOUT_FILENO);
+    dpdk_helper_run();
```

Comment:
Please add an `exit()` or `abort()` after this line to ensure that `dpdk_helper_spawn()` does not return twice.